jvm-packages/xgboost4j-spark-gpu/src/main/scala/ml/dmlc/xgboost4j/scala/spark/ExternalMemory.scala

```diff
@@ -0,0 +1,220 @@
+  }
+
+  /**
+   * Load the path from disk to the Table
+   *
+   * @param name to be loaded
+   * @return Table
+   */
+  override def loadTable(name: String): Table = {
+    val file = new File(name)
+    if (!file.exists()) {
+      throw new RuntimeException(s"The cache file ${name} doesn't exist" )
+    }
+    try {
+      withResource(Table.readArrowIPCChunked(file)) { reader =>
+        val tables = ArrayBuffer.empty[Table]
+        closeOnExcept(tables) { tables =>
+          var table = Option(reader.getNextIfAvailable())
+          while (table.isDefined) {
+            tables.append(table.get)
+            table = Option(reader.getNextIfAvailable())
+          }
+        }
+        if (tables.size > 1) {
+          closeOnExcept(tables) { tables =>
+            Table.concatenate(tables.toArray: _*)
+          }
+        } else {
+          tables(0)
+        }
+      }
+    } catch {
+      case e: Throwable =>
+        close()
+        throw e
+    } finally {
+      if (file.exists()) {
+        file.delete()
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    buffers.foreach { path =>
+      val file = new File(path)
+      if (file.exists()) {
+        file.delete()
+      }
+    }
+    buffers.clear()
+  }
+}
+
+private[spark] object ExternalMemory {
+  def apply(path: Option[String] = None): ExternalMemory[_] = {
+    path.map(new DiskExternalMemoryIterator(_))
+      .getOrElse(throw new RuntimeException("No disk path provided"))
+  }
+}
+
+/**
+ * ExternalMemoryIterator supports iterating the data twice if the `swap` is called.
+ *
+ * The first round iteration gets the input batch that will be
+ *   1. cached in the external memory
+ *      2. fed in QuantilDmatrix
+ *      The second round iteration returns the cached batch got from external memory.
+ *
+ * @param input   the spark input iterator
+ * @param indices column index
+ */
+private[scala] class ExternalMemoryIterator(val input: Iterator[Table],
+                                            val indices: ColumnIndices,
+                                            val path: Option[String] = None)
+  extends Iterator[ColumnBatch] {
+
+  private var iter = input
+
+  // Flag to indicate the input has been consumed.
+  private var inputIsConsumed = false
+  // Flag to indicate the input.next has been called which is valid
+  private var inputNextIsCalled = false
+
+  // visible for testing
+  private[spark] val externalMemory = ExternalMemory(path)
+
+  override def hasNext: Boolean = {
+    val value = iter.hasNext
+    if (!value && inputIsConsumed && inputNextIsCalled) {
+      externalMemory.close()
+    }
+    if (!inputIsConsumed && !value && inputNextIsCalled) {
+      inputIsConsumed = true
+      iter = externalMemory
+    }
+    value
+  }
+
+  override def next(): ColumnBatch = {
+    inputNextIsCalled = true
+    withResource(new GpuColumnBatch(iter.next())) { batch =>
+      if (iter.eq(input)) {
+        externalMemory.cacheTable(batch.table)
```

Comment:
It would be great if we could write data asynchronously in the future (after this PR). This way, we can let XGBoost handle the batch while it's being written simultaneously.